test/Hamiltonians.jl

```diff
@@ -908,6 +908,10 @@ end
     @test diagonal_element(S0, uniform_state) == 0
     @test diagonal_element(S2, uniform_state) == 0
 
+    # Test return type for integer, and non-integer filling
+    @test diagonal_element(S0, localised_state) isa Float64
+    @test diagonal_element(S1, non_unital_localised_state) isa ComplexF64
```

Comment:
While you have tested that the functions return the type you are expecting for the examples, there is still a type instability (which means that the compiler cannot figure out which type will need to be returned). It is better to test that the the `@inferred` macro:
```julia-repl
julia> @inferred diagonal_element(S0, localised_state)
5.0

julia> @inferred diagonal_element(S1, non_unital_localised_state)
ERROR: return type ComplexF64 does not match inferred return type Any
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] top-level scope
   @ REPL[13]:1
```
A good way of getting more information is using the `@code_warntype` macro.

The type instability has consequences for the execution time:
```julia-repl
julia> using BenchmarkTools

julia> @btime diagonal_element($S0, $localised_state)
  2.791 ns (0 allocations: 0 bytes)
5.0

julia> @btime diagonal_element($S1, $non_unital_localised_state)
  1.179 μs (44 allocations: 1.08 KiB)
0.22222222222222213 + 0.3849001794597506im
```
You see that the type-unstable (complex) code is almost a factor 1000 times slower and allocates memory. The fact that the compiler cannot figure out the return type can also contaminate the rest of your code (handling the results). So you really want to avoid this type instability.

Honestly, looking through the code I don't really understand where the type instability comes from. Maybe @mtsch can help?
```julia-repl
julia> @code_warntype diagonal_element(S1, non_unital_localised_state)
MethodInstance for Rimu.Interfaces.diagonal_element(::StringCorrelator{1}, ::BoseFS{4, 6, BitString{9, 1, UInt16}})
  from diagonal_element(::StringCorrelator{D}, add::SingleComponentFockAddress) where D @ Rimu.Hamiltonians ~/git/code/Rimu/src/Hamiltonians/correlation_functions.jl:347
Static Parameters
  D = 1
Arguments
  #self#::Core.Const(Rimu.Interfaces.diagonal_element)
  _::Core.Const(StringCorrelator(1))
  add::BoseFS{4, 6, BitString{9, 1, UInt16}}
Locals
  @_4::Union{Nothing, Tuple{Int64, Int64}}
  @_5::Union{}
  result::Any
  n̄@_7::Core.Box
  v::StaticArraysCore.SVector{6, Int32}
  d::Int64
  N::Int64
  M::Int64
  #81::Union{}
  i@_13::Union{}
  phase_sum@_14::Union{}
  #82::Rimu.Hamiltonians.var"#82#84"{StaticArraysCore.SVector{6, Int32}, Int64}
  i@_16::Int64
  phase_sum@_17::Any
  n̄@_18::Union{}
  n̄@_19::Union{}
  n̄@_20::Union{}
  n̄@_21::Union{}
Body::Any
1 ──        Core.NewvarNode(:(@_4))
│           Core.NewvarNode(:(@_5))
│           Core.NewvarNode(:(result))
│           (n̄@_7 = Core.Box())
│           (M = Rimu.Hamiltonians.num_modes(add))
│           (N = Rimu.Hamiltonians.num_particles(add))
│           (d = Rimu.Hamiltonians.mod($(Expr(:static_parameter, 1)), M::Core.Const(6)))
│           (v = Rimu.Hamiltonians.onr(add))
│    %9   = (N::Core.Const(4) % M::Core.Const(6))::Core.Const(4)
│    %10  = Rimu.Hamiltonians.iszero(%9)::Core.Const(false)
└───        goto #3 if not %10
2 ──        Core.Const(:(N ÷ M))
│           Core.Const(:(Core.setfield!(n̄@_7, :contents, %12)))
│           Core.Const(:(result = 0.0))
│           Core.Const(:(Rimu.Hamiltonians.eachindex(v)))
│           Core.Const(:(@_5 = Base.iterate(%15)))
│           Core.Const(:(@_5 === nothing))
│           Core.Const(:(Base.not_int(%17)))
│           Core.Const(:(goto %65 if not %18))
│           Core.Const(:(@_5))
│           Core.Const(:(i@_13 = Core.getfield(%20, 1)))
│           Core.Const(:(Core.getfield(%20, 2)))
│           Core.Const(:(Rimu.Hamiltonians.:(var"#81#83")))
│           Core.Const(:(Core.typeof(v)))
│           Core.Const(:(Core.typeof(M)))
│           Core.Const(:(Core.apply_type(%23, %24, %25)))
│           Core.Const(:(n̄@_7))
│           Core.Const(:(v))
│           Core.Const(:(#81 = %new(%26, %27, %28, M)))
│           Core.Const(:(#81))
│           Core.Const(:(i@_13))
│           Core.Const(:(i@_13 + d))
│           Core.Const(:(%32 - 1))
│           Core.Const(:(%31:1:%33))
│           Core.Const(:(Base.Generator(%30, %34)))
│           Core.Const(:(phase_sum@_14 = Rimu.Hamiltonians.sum(%35)))
│           Core.Const(:(result))
│           Core.Const(:(Base.getindex(v, i@_13)))
│           Core.Const(:(Core.isdefined(n̄@_7, :contents)))
│           Core.Const(:(goto %42 if not %39))
│           Core.Const(:(goto %44))
│           Core.Const(Core.NewvarNode(:(n̄@_18)))
│           Core.Const(:(n̄@_18))
│           Core.Const(:(Core.getfield(n̄@_7, :contents)))
│           Core.Const(:(%38 - %44))
│           Core.Const(:((-1) ^ phase_sum@_14))
│           Core.Const(:(v))
│           Core.Const(:(i@_13 + d))
│           Core.Const(:(Rimu.Hamiltonians.mod1(%48, M)))
│           Core.Const(:(Base.getindex(%47, %49)))
│           Core.Const(:(Core.isdefined(n̄@_7, :contents)))
│           Core.Const(:(goto %54 if not %51))
│           Core.Const(:(goto %56))
│           Core.Const(Core.NewvarNode(:(n̄@_19)))
│           Core.Const(:(n̄@_19))
│           Core.Const(:(Core.getfield(n̄@_7, :contents)))
│           Core.Const(:(%50 - %56))
│           Core.Const(:(%45 * %46 * %57))
│           Core.Const(:(result = %37 + %58))
│           Core.Const(:(@_5 = Base.iterate(%15, %22)))
│           Core.Const(:(@_5 === nothing))
│           Core.Const(:(Base.not_int(%61)))
│           Core.Const(:(goto %65 if not %62))
│           Core.Const(:(goto %20))
│           Core.Const(:(result / M))
└───        Core.Const(:(return %65))
3 ┄─ %67  = (N::Core.Const(4) / M::Core.Const(6))::Core.Const(0.6666666666666666)
│           Core.setfield!(n̄@_7, :contents, %67)
│           (result = Rimu.Hamiltonians.ComplexF64(0))
│    %70  = Rimu.Hamiltonians.eachindex(v)::Core.Const(SOneTo(6))
│           (@_4 = Base.iterate(%70))
│    %72  = (@_4::Core.Const((1, 1)) === nothing)::Core.Const(false)
│    %73  = Base.not_int(%72)::Core.Const(true)
└───        goto #12 if not %73
4 ┄─ %75  = @_4::Tuple{Int64, Int64}
│           (i@_16 = Core.getfield(%75, 1))
│    %77  = Core.getfield(%75, 2)::Int64
│    %78  = Rimu.Hamiltonians.:(var"#82#84")::Core.Const(Rimu.Hamiltonians.var"#82#84")
│    %79  = Core.typeof(v)::Core.Const(StaticArraysCore.SVector{6, Int32})
│    %80  = Core.typeof(M::Core.Const(6))::Core.Const(Int64)
│    %81  = Core.apply_type(%78, %79, %80)::Core.Const(Rimu.Hamiltonians.var"#82#84"{StaticArraysCore.SVector{6, Int32}, Int64})
│    %82  = n̄@_7::Core.Box
│    %83  = v::StaticArraysCore.SVector{6, Int32}
│           (#82 = %new(%81, %82, %83, M::Core.Const(6)))
│    %85  = #82::Core.PartialStruct(Rimu.Hamiltonians.var"#82#84"{StaticArraysCore.SVector{6, Int32}, Int64}, Any[Core.Box, StaticArraysCore.SVector{6, Int32}, Core.Const(6)])
│    %86  = i@_16::Int64
│    %87  = (i@_16 + d::Core.Const(1))::Int64
│    %88  = (%87 - 1)::Int64
│    %89  = (%86:1:%88)::Core.PartialStruct(StepRange{Int64, Int64}, Any[Int64, Core.Const(1), Int64])
│    %90  = Base.Generator(%85, %89)::Core.PartialStruct(Base.Generator{StepRange{Int64, Int64}, Rimu.Hamiltonians.var"#82#84"{StaticArraysCore.SVector{6, Int32}, Int64}}, Any[Core.PartialStruct(Rimu.Hamiltonians.var"#82#84"{StaticArraysCore.SVector{6, Int32}, Int64}, Any[Core.Box, StaticArraysCore.SVector{6, Int32}, Core.Const(6)]), StepRange{Int64, Int64}])
│           (phase_sum@_17 = Rimu.Hamiltonians.sum(%90))
│    %92  = result::Any
│    %93  = Base.getindex(v, i@_16)::Int32
│    %94  = Core.isdefined(n̄@_7, :contents)::Bool
└───        goto #6 if not %94
5 ──        goto #7
6 ──        Core.NewvarNode(:(n̄@_20))
└───        n̄@_20
7 ┄─ %99  = Core.getfield(n̄@_7, :contents)::Any
│    %100 = (%93 - %99)::Any
│    %101 = (Rimu.Hamiltonians.pi * Rimu.Hamiltonians.im * phase_sum@_17)::Any
│    %102 = Rimu.Hamiltonians.exp(%101)::Any
│    %103 = v::StaticArraysCore.SVector{6, Int32}
│    %104 = (i@_16 + d::Core.Const(1))::Int64
│    %105 = Rimu.Hamiltonians.mod1(%104, M::Core.Const(6))::Int64
│    %106 = Base.getindex(%103, %105)::Int32
│    %107 = Core.isdefined(n̄@_7, :contents)::Bool
└───        goto #9 if not %107
8 ──        goto #10
9 ──        Core.NewvarNode(:(n̄@_21))
└───        n̄@_21
10 ┄ %112 = Core.getfield(n̄@_7, :contents)::Any
│    %113 = (%106 - %112)::Any
│    %114 = (%100 * %102 * %113)::Any
│           (result = %92 + %114)
│           (@_4 = Base.iterate(%70, %77))
│    %117 = (@_4 === nothing)::Bool
│    %118 = Base.not_int(%117)::Bool
└───        goto #12 if not %118
11 ─        goto #4
12 ┄ %121 = (result / M::Core.Const(6))::Any
└───        return %121
```

Reply:
Nice. The code change helped:
```julia-repl
julia> @btime diagonal_element($S1, $non_unital_localised_state)
  99.709 ns (0 allocations: 0 bytes)
0.22222222222222213 + 0.3849001794597506im

julia> @btime diagonal_element($S0, $localised_state)
  2.791 ns (0 allocations: 0 bytes)
5.0

julia> @btime diagonal_element($S1, $localised_state)
  31.983 ns (0 allocations: 0 bytes)
1.0
```
The non-complex calculation is a factor of three faster than using complex numbers, but there are no allocations and time scales are in the ns regime.